File: android-compose-example/app/src/main/java/net/irext/ircontrol/compose/controller/PhoneRemote.java
package net.irext.ircontrol.compose.controller;

import android.content.Context;
import android.util.Log;
import net.irext.decode.sdk.IRDecode;
import net.irext.decode.sdk.bean.ACStatus;
import net.irext.ircontrol.compose.controller.base.ControlHelper;
import net.irext.ircontrol.compose.controller.base.Remote;

/**
 * Filename:       PhoneRemote.java
 * Revised:        Date: 2026-01-18
 * Revision:       Revision: 1.0
 * <p>
 * Description:    Remote implementation by Android phone
 * <p>
 * Revision log:
 *2026-01-18: created by strawmanbobi
 */
public class PhoneRemote extends Remote {

    private static final String TAG = PhoneRemote.class.getSimpleName();

    private Context mContext;
    private IRDecode mIRDecode;

    private static PhoneRemote mInstance;

    public PhoneRemote(Context context) {
        mContext = context;
        mIRDecode = IRDecode.getInstance();
    }

    public static PhoneRemote getInstance(Context context) {
        if (mInstance == null) {
            mInstance = new PhoneRemote(context);
        }
        return mInstance;
    }

    public int irOpen(String remoteBinFilePath, int category, int subCategory) {
        return mIRDecode.openFile(category, subCategory, remoteBinFilePath);
    }

    public int irControl(int category, int subCategory, int keyCode) {
        int []decoded;
        StringBuilder debugStr = new StringBuilder();
        ACStatus acStatus = new ACStatus();
        int inputKeyCode = ControlHelper.translateKeyCode(category, keyCode, acStatus);
        decoded = mIRDecode.decodeBinary(inputKeyCode, acStatus);

        for (int i = 0; i < decoded.length; i++) {
            debugStr.append(decoded[i]);
            if (i != decoded.length - 1) {
                debugStr.append(",");
            }
        }
        Log.d(TAG, "IR control decoded: " + debugStr);
        ControlHelper.transmitIr(mContext, decoded);
        return 0;
    }

    public void irClose() {
        mIRDecode.closeBinary();
    }
}

File: android-compose-example/app/src/main/java/net/irext/ircontrol/compose/controller/PhoneRemote.kt
package net.irext.ircontrol.compose.controller

import android.content.Context
import android.util.Log
import net.irext.decode.sdk.IRDecode
import net.irext.decode.sdk.bean.ACStatus

/**
 * Filename:       PhoneRemote.kt
 * Created:        Date: 2026-07-09
 *
 * Description:    Sends decoded IR commands through the phone IR emitter.
 *
 * Revision log:
 * 2026-07-09: created by shdmfire and strawmanbobi
 */

class PhoneRemote(
    context: Context,
    private val irTransmitter: IrTransmitter = IrTransmitter(context),
) {

    private val irDecode: IRDecode = IRDecode.getInstance()

    fun openBinary(remoteBinFilePath: String, category: Int, subCategory: Int): Int {
        return irDecode.openFile(category, subCategory, remoteBinFilePath)
    }

    fun control(category: Int, subCategory: Int, command: ControlCommand): ControlResult {
        val acStatus = ACStatus()
        val inputKeyCode = command.toDecodeKeyCode(category, acStatus)
        val decoded = irDecode.decodeBinary(inputKeyCode, acStatus) ?: return ControlResult.Failed

        Log.d(TAG, "IR control decoded: ${decoded.joinToString(",")}")
        return if (irTransmitter.transmit(decoded)) ControlResult.Success else ControlResult.Failed
    }

    fun closeBinary() {
        irDecode.closeBinary()
    }

    private companion object {
        private val TAG = PhoneRemote::class.java.simpleName
    }
}
